tests/fixtures.hh

```diff
@@ -0,0 +1,259 @@
+        if(tree) {
+            tree->insert_root();
+            root = tree->root();
+        }
+    }
+
+    virtual ~TrieFixture() {
+//        TODO: This causes segfaults in the tree's desctrutor when deleting a trivial tree
+//        without any nodes except the root node. Most likely, the bug occurs because 
+//        the root node gets deleted, but its reference in the Tree class is not set to null.
+//        Thus in the tree's destructor, the tree thinks it has a root node when it doesn't
+//        and tries to free already freed memory.
+//
+        if(tree)
+            delete tree;
+    }
+
+protected:
+    double c;
+    const char * type;
+    int ablation;
+    bool calculate_size;
+    CacheTree * tree;
+    Node * root;
+};
+
+
+class PrefixMapFixture {
+
+public:
+    PrefixMapFixture() : pmap(NULL), tree(NULL), root(NULL), num_not_captured(T_NNC),
+                         c(T_C), prediction(T_PRED), default_prediction(T_DEF_PRED),
+                         lower_bound(T_LOWER), objective(T_OBJ), len_prefix(4), equivalent_minority(T_MINOR) {
+        pmap = new PrefixPermutationMap();
+
+        tree = new CacheTree(T_NSAMPLES, T_NRULES, T_C, rules, labels, NULL, T_ABLATION, false, T_TYPE);
+
+        if(tree) {
+            tree->insert_root();
+            root = tree->root();
+        }
+
+        // Canonical prefix, also the prefix map key
+        correct_key = {4, 1, 2, 4, 5};
+
+        // First insertion
+        parent_prefix = {4, 2, 1};
+        new_rule = 5;
+        correct_indices = {4, 2, 1, 0, 3};
+
+        // Second insertion, for second and third tests
+        parent_prefix_2 = {1, 4, 5};
+        new_rule_2 = 2;
+        correct_indices_2 = {4, 0, 3, 1, 2};
+    }
+
+    virtual ~PrefixMapFixture() {
+
+        if(pmap)
+            delete pmap;
+
+        if(tree)
+            delete tree;
+    }
+
+protected:
+
+    PrefixPermutationMap * pmap;
+    CacheTree * tree;
+    Node * root;
+
+    int num_not_captured;
+
+    double c;
+
+    bool prediction;
+    bool default_prediction;
+    double lower_bound;
+    double objective;
+    int len_prefix;
+
+    std::vector<unsigned short> correct_key;
+
+    tracking_vector<unsigned short, DataStruct::Tree> parent_prefix;
+    int new_rule;
+    std::vector<unsigned char> correct_indices;
+
+    tracking_vector<unsigned short, DataStruct::Tree> parent_prefix_2;
+    int new_rule_2;
+    std::vector<unsigned char> correct_indices_2;
+
+    double equivalent_minority;
+};
+
+
+class CapturedMapFixture {
+
+public:
+    CapturedMapFixture() : pmap(NULL), tree(NULL), root(NULL), num_not_captured(T_NNC),
+                         c(T_C), prediction(T_PRED), default_prediction(T_DEF_PRED),
+                         lower_bound(T_LOWER), objective(T_OBJ), len_prefix(4), equivalent_minority(T_MINOR) {
+        pmap = new CapturedPermutationMap();
+
+        tree = new CacheTree(T_NSAMPLES, T_NRULES, T_C, rules, labels, NULL, T_ABLATION, false, T_TYPE);
```

Comment:
Why are you creating a tree here?

Reply:
Because when we want to insert into the permutation map, which is the case for all tests with this fixture, we also need to supply a tree to the pmap->insert function.